src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -80,19 +93,136 @@ public void completed(Integer result, Void v) {
+                                            file.getAbsolutePath());
+                                    deleteFile(file);
+                                    callback.onPutComplete();
+                                    return;
+                                }
+                                header.size = resultLen;
+                                mTotalSize =
+                                        DiskBasedCacheUtility.putEntry(
+                                                key, header, mTotalSize, mEntries);
+                                mTotalSize =
+                                        DiskBasedCacheUtility.pruneIfNeeded(
+                                                mTotalSize,
+                                                mMaxCacheSizeInBytes,
+                                                mEntries,
+                                                mRootDirectorySupplier);
+                            } else {
+                                deleteFile(file);
+                            }
+
+                            callback.onPutComplete();
+                        }
+
+                        @Override
+                        public void failed(Throwable throwable, Void ignore) {
+                            VolleyLog.e(
+                                    throwable, "Failed to read file %s", file.getAbsolutePath());
+                            deleteFile(file);
+                            callback.onPutComplete();
+                            closeChannel(afc, "failed read");
+                        }
+                    });
+        } catch (IOException e) {
+            if (closeChannel(channel, "IOException")) {
+                deleteFile(file);
+            }
+            callback.onPutComplete();
+        }
+    }
+
+    public void initialize() {
+        // TODO (sphill99): #181
+    }
+
+    /** Re-initialize the cache if the directory was deleted. */
+    private void initializeIfRootDirectoryDeleted() {
+        if (mRootDirectorySupplier.get().exists()) {
+            return;
+        }
+        VolleyLog.d("Re-initializing cache after external clearing.");
+        mEntries.clear();
+        mTotalSize = 0;
+        initialize();
+    }
+
+    /**
+     * Closes the asynchronous file channel.
+     *
+     * @param afc Channel that is being closed.
+     * @param endOfMessage End of error message that logs where the close is happening.
+     * @return Returns true if the channel is successfully closed, false if channel is null, or
+     *     close results in an IOException.
+     */
+    private boolean closeChannel(@Nullable AsynchronousFileChannel afc, String endOfMessage) {
+        if (afc == null) {
+            return false;
+        }
+        try {
+            afc.close();
+            return true;
+        } catch (IOException e) {
+            VolleyLog.e(e, "failed to close file after " + endOfMessage);
```

Comment:
nit: use the String format type of log: `log(e. "failed to close file after %s", endOfMessage)`